### ISSUE TITLE
snapdrop.net broken with cookies enabled

### ISSUE THREAD

**Issue by SebiderSushi** [bromite/bromite#1693]
### Bromite version

Version: 96.0.4664.54
Arch: `arm`
Android version: Android 11
Device model: OnePlus X

### Flags

No flags were changed under `chrome://flags`.

### Is this bug about the SystemWebView?

Yes (both WebView and Bromite browser are affected)

### Is the bug reproducible with latest version?

Yes

### Can the bug be reproduced with corresponding Chromium version?

No

### Is this bug happening in an incognito tab?

Yes

### Allow ads for the site; is the bug caused by the adblocker?

No

### Is the bug a crash?

No

### Describe the bug

When initially loading the page snapdrop.net it resolves a device name from the user agent, displays the random advertisal name at the bottom and scans for other devices on the network and displays them.
When reloading the page / restarting Bromite and revisiting it none of these things happen. A default string remains at the bottom of the screen, no other devices are found and the device bromite is running on is not found by other browsers on the same network.

### Steps to reproduce the bug

Steps to reproduce the bug:
1. Go to https://snapdrop.net
2. Reload the page
3. Check bottom of the screen
5. Tap on the Lock in the URL bar, then Cookies. Delete the cookies.
6. Reload - Page works as expected
7. Reload again - Error reappears
8. Tap on the Lock in the URL bar, then Permissions. Disable Cookies.
9. Reload - Page works as expected

The bug can also be reproduced in webview based browsers or apps that display the page snapdrop.net:
1. Install https://f-droid.org/en/packages/com.fmsys.snapdrop/
2. Launch - works as expected
3. Reload using the pull down to refresh gesture - Bug reproduced
4. Go to App Info and clear app data for Snapdrop
5. Relaunch - works as expected until page reload / app restart

### Expected behavior

snapdrop.net successfully starts up, discovers other devices and is discoverable without the need to change preferences as a workaround. WebView based clients with no feasable workarounds work as well.

**Comment by kokukyocho** [comment]
Hello @SebiderSushi!

I am a bot :robot:

Thanks for submitting this issue!
I noticed that it is missing the template, please edit the issue to match the template for either a [bug](https://github.com/bromite/bromite/blob/master/.github/ISSUE_TEMPLATE/bug_report.md) or a [feature request](https://github.com/bromite/bromite/blob/master/.github/ISSUE_TEMPLATE/feature_request.md); the issue will be automatically re-opened afterwards.

If this is not an issue nor a feature request and you want to start a discussion then you can use [GitHub Discussions](https://github.com/bromite/bromite/discussions).

**Tip**: if you use GitHub in Desktop mode the template will be proposed to you when submitting issues.

**Comment by SebiderSushi** [comment]
Sorry! I fail to understand why this bot does not detect the bug issue template in my issue. I'd readily fix it but i have absolutely no idea what the bot wants me to do. When creating the issue i started out by choosing the bug template, then filled everything in.

**Comment by Mr-Administrator** [comment]
Evidently, you're not alone, friend...
https://github.com/bromite/bromite/issues/1695

**Comment by csagan5** [comment]
@SebiderSushi the bug reports can only be about a single architecture and Android version

**Comment by uazo** [comment]
I confirm, it is reproducible.
this is the error
```
network.js:16 Uncaught DOMException: Failed to construct 'WebSocket': An insecure WebSocket connection may not be initiated from a page loaded over HTTPS.
    at ServerConnection._connect (https://snapdrop.net/scripts/network.js:16:20)
    at new ServerConnection (https://snapdrop.net/scripts/network.js:7:14)
    at new Snapdrop (https://snapdrop.net/scripts/ui.js:523:24)
    at https://snapdrop.net/scripts/ui.js:538:18
```
although the connection is towards `wss://snapdrop.net/server/webrtc`
the bug is in `Block-gateway-attacks-via-websockets.patch` at line
```
third_party/blink/renderer/core/loader/frame_fetch_context.cc

  if (GetBaseFetchContext()->ShouldBlockGateWayAttacks(execution_context_->AddressSpace(), url)) {
    has_initiated_opening_handshake_ = false;
    return false;
  }
```
because `execution_context_->AddressSpace()` is equals to `kUnknown`
cleaning the cookies instead you have `kPublic`, maybe the service worker has something to do with it.
we need to handle `kUnknown` as public, as indicated

```
  kUnknown,  // For security checks, "unknown" will be treated as "public", as
             // that's the lowest-privilege value.
```
so in `FrameFetchContext::ShouldBlockGateWayAttacks` we need to add an if

**Comment by uazo** [comment]
[this](https://github.com/uazo/bromite/blob/master-v97/build/patches/Block-gateway-attacks-via-websockets.patch) should solve

**Comment by csagan5** [comment]
> [this](https://github.com/uazo/bromite/blob/master-v97/build/patches/Block-gateway-attacks-via-websockets.patch) should solve

That's great, are you going to make a PR for it?

In the past it was never possible to understand where the bug was, I am glad you could identify it; since the original code was from upstream it is possible that Chromium would also benefit from a similar patch.

Related issues:
* #1373
* #1320
* #1294
* #1189
* #1129

**Comment by uazo** [comment]
> That's great, are you going to make a PR for it?

sure, I was waiting for you to come back to life :)
I have stuff to merge ..,

**Comment by SebiderSushi** [comment]
Thanks a lot for the fix @uazo

**Comment by csagan5** [comment]
Fixed in `98.0.4758.108`.

**Comment by SebiderSushi** [comment]
Thanks for the update! I just downloaded and tried [v98.0.4758.108](https://github.com/bromite/bromite/releases/tag/98.0.4758.108) but unfortunately, i am still able to reproduce the issue by following the reproduction steps. These are the scenarios i tested:
 - Updated Bromite 64bit on my daily driver without wiping any app data
 - Updated the Bromite WebView 64bit on my daily driver and tested with the WebView based browser Jelly from LineageOS without wiping any data
 - Updated the Bromite WebView 64bit on my daily driver and tested with a fresh install of https://f-droid.org/en/packages/com.fmsys.snapdrop/
 - Fresh install of Bromite 32bit on a separate device

**Comment by uazo** [comment]
I confirm the non-resolution, although I do not yet understand the reason. I'll handle that.

**Comment by uazo** [comment]
my fault, there were 3 `ShouldBlockGateWayAttacks` to be changed ...

forgot the `WebSocketCommon::ShouldBlockGateWayAttacks`, @csagan5 do you take care of it?

**Comment by csagan5** [comment]
I see you already made a PR; it is now merged so it should be fixed on next release.

**Comment by SebiderSushi** [comment]
thanks again for the quick responses and thanks for fixing, the bug is no longer reproducable.

unfortunately, now that i no longer wasn't focused on this specific bug i found out that most of the main functionality like sending and receiving files does not work in bromite while it does in vanilla chromium, at least with another chromium browser as the sender/recipient, that is to say i also wasn't able to make it work between vanilla chromium / LineageOS Chromium WebView on my mobile and a fresh profile of firefox on desktop, which would be my main use case. given this flaky experience i had with snapdrop i will most likely just move on to other file transfer solutions since guess i lost my pacience this time around.

**TL;DR:** feel free when handling this specific issue, i guess the initial bug is fixed but snapdrop does not work with Bromite as it does with vanilla Chromium. If anyone else is interested in making snapdrop work with Bromite feel free to investigate.

**Comment by wallismith** [comment]
I was following this issue and was hoping this issue would also fix the issue I'm having. Unlike @SebiderSushi , I'm running snapdrop on my local network and, as he mentions, sending and receiving files don't work when using Bromite. I've tried Firefox, Brave/Chrome and am able to send and receive files. I also looked in the browser console when using Bromite to see if the WebSocket failed to connect, but it successfully connects and pings.

Devices I tried:
Bromite version: 98.0.4758.116
Android version: Android 12 (GrapheneOS)
Device model: Pixel 3

and

Bromite version: 98.0.4758.116
Android version: Android 12
Device model: Pixel 6

Feel free to reach out if you have questions. Thanks.

**Comment by csagan5** [comment]
Fixed in `98.0.4758.116`.

**Comment by fm-sys** [comment]
> Yeah! We can detect other devices as expected! BUT, we still cannot send and receive files or even messages...
This issue is just partially fixed...

_Originally posted by @DodoLeDev in https://github.com/fm-sys/snapdrop-android/issues/163#issuecomment-1059811134_

So this issue is only fixed partially. File transfers do still not work - maybe cause WebRTC is blocked...?

**Comment by csagan5** [comment]
@fm-sys the original issue is solved; feel free to open a new issue **after** figuring out what the issue is.

**Comment by fm-sys** [comment]
The original issue was, that snapdrop.net is not working with bromite. And this is still 100% valid, so the problem isn't solved in any way.

Side note:
I'm the developer of the Snapdrop android app, but not using bromite myself. I will definitely not have the time to debug this in my spare time. Either someone using Snapdrop and Bromite is interested in diving into this further, or Snapdrop users will have to say goodbye to bromite. And that would be sad...

**Comment by csagan5** [comment]
> The original issue was, that snapdrop.net is not working with bromite. And this is still 100% valid, so the problem isn't solved in any way.

The issue submitter is expected to figure out what is not working and submit an issue for that; submitting a minimal testcase that does not involve using a specific app or website is necessary for this.

> Snapdrop users will have to say goodbye to bromite. And that would be sad...

Not sad at all. It would be sad if we would open an issue for every website that does not work with Bromite instead of submitting an issue for the actual broken API or functionality, and even before that checking whether it is working as intended or not.
